### PR TITLE
BUG: Fix various problems with the np.ma.masked constant

### LINKED ISSUE #9328: BUG: MaskedConstant  is mutated despite copying
Copying the masked constant shouldn't be possible, and definitely shouldn't allow this to happen
```python
# ok, this is sane
>>> np.ma.masked[()] = 1
MaskError: Cannot alter the masked element.

# making a copy allows modification - seems reasonable
>>> x = np.ma.masked.copy()
>>> x[()] = 2
>>> x.data
array( 2.)

# But the repr is pretty  confusing
>>> x
masked
>>> x is np.ma.masked
False

# And somehow our data is lost by casting?
>>> x.view(np.ma.MaskedArray)
masked_array(data = 0.0,
             mask = False,
       fill_value = 1e+20)

# Worse yet, np.ma.masked is no longer masked!
>>> np.ma.masked.view(np.ma.MaskedArray)
masked_array(data = 2.0,
             mask = False,
       fill_value = 1e+20)
```

So now we can hide arbitrary values inside a fake masked constant

### PR DESCRIPTION
**Pull request by eric-wieser**
It turns out that `np.ma.masked` is actually pretty bad at preserving identity or value

This PR attempts to fix that.

Fixes #9328 and #4595

### CODE REVIEW

File: numpy/ma/core.py [hunk 1]
```diff
@@ -6184,32 +6184,43 @@ def isMaskedArray(x):
...
+            # properties normally
+            return super(MaskedConstant, self).__array_finalize__(obj)
+        elif self is self.__singleton:
+            # not clear how this can happen, play it safe
+            pass
+        else:
+            # everywhere else, we want to downcast to MaskedArray, to prevent a
+            # duplicate maskedconstant.
+            self.__class__ = MaskedArray
+            self.__array_finalize__(obj)
```
Comment (eric-wieser) [inline numpy/ma/core.py]: I can't see a good alternative here to assigning to `__class__` - unfortunately, `__array_finalize__` asks us to operate in place, rather than return a new object.

Does this work on PyPy, @mattip?

File: numpy/ma/tests/test_core.py [hunk 2]
```diff
@@ -4732,6 +4732,40 @@ def test_ctor(self):
...
+
+        # original should be unchanged
+        assert_equal(np.ma.masked.data, 0)
+        assert_equal(np.ma.masked.mask, True)
+
+    def test_no_copies(self):
+        MC = np.ma.core.MaskedConstant
+        assert_(not isinstance(np.ma.masked[...], MC))
+        assert_(not isinstance(np.ma.masked.copy(), MC))
+
```
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: All of these tests previously failed

File: numpy/ma/core.py (hunk 1 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Seems to be alright on `PyPy.js`, with a fake inheritance of `MaskedConstant -> MaskedArray -> int`

**Comment by charris** [comment]
@mhvk @ahaldane Could you folks take a look at this?

File: numpy/ma/core.py [hunk 3]
```diff
@@ -6184,17 +6184,38 @@ def isMaskedArray(x):
 
 
 class MaskedConstant(MaskedArray):
-    # We define the masked singleton as a float for higher precedence.
-    # Note that it can be tricky sometimes w/ type comparison
-    _data = data = np.array(0.)
-    _mask = mask = np.array(True)
-    _baseclass = ndarray
+
+    __singleton = None
```
Comment (mhvk) [inline numpy/ma/core.py]: Why not just a single `_`?

File: numpy/ma/core.py [hunk 4]
```diff
@@ -6184,17 +6184,38 @@ def isMaskedArray(x):
...
 
     def __new__(self):
```
Comment (mhvk) [inline numpy/ma/core.py]: While you're at it, could you change `self` to `cls` here - this really is super confusing.

File: numpy/ma/core.py [hunk 5]
```diff
@@ -6184,17 +6184,38 @@ def isMaskedArray(x):
...
+            # Note that it can be tricky sometimes w/ type comparison
+            data = np.array(0.)
+            mask = np.array(True)
+
+            # prevent any modifications
+            data.flags.writeable = False
+            mask.flags.writeable = False
+
+            masked = MaskedArray(data, mask=mask)
+            self.__singleton = masked.view(self)
```
Comment (mhvk) [inline numpy/ma/core.py]: This especially would be nicer as `cls._singleton = ...`

File: numpy/ma/core.py [hunk 6]
```diff
@@ -6184,17 +6184,38 @@ def isMaskedArray(x):
...
+            # We define the masked singleton as a float for higher precedence.
+            # Note that it can be tricky sometimes w/ type comparison
+            data = np.array(0.)
+            mask = np.array(True)
+
+            # prevent any modifications
+            data.flags.writeable = False
+            mask.flags.writeable = False
+
+            masked = MaskedArray(data, mask=mask)
```
Comment (mhvk) [inline numpy/ma/core.py]: Here, I'd avoid the view casing and do instead (see comment on `__array_finalize__`):
```
masked = super(MaskedConstant, cls).__new__(cls, data, mask=mask)
```

File: numpy/ma/core.py [hunk 7]
```diff
@@ -6184,17 +6184,38 @@ def isMaskedArray(x):
...
+            self.__singleton = masked.view(self)
+
+        return self.__singleton
 
     def __array_finalize__(self, obj):
-        return
+        if self.__singleton is None:
```
Comment (mhvk) [inline numpy/ma/core.py]: Wouldn't it make more sense to choose path based on `obj`? If one uses `super` in `__new__` as suggested above, then the relevant case that is OK is `obj is None` - all other cases imply view casting or slicing, neither of which I think should be allowed (except if `obj is masked`) - could they just raise?

File: numpy/ma/core.py [hunk 8]
```diff
@@ -6206,16 +6227,23 @@ def __str__(self):
         return str(masked_print_option._display)
 
     def __repr__(self):
-        return 'masked'
-
-    def flatten(self):
```
Comment (mhvk) [inline numpy/ma/core.py]: Why is this removed?

File: numpy/ma/core.py [hunk 9]
```diff
@@ -6206,16 +6227,23 @@ def __str__(self):
...
+            return object.__repr__(self)
 
     def __reduce__(self):
         """Override of MaskedArray's __reduce__.
         """
         return (self.__class__, ())
 
+    # inplace operations have no effect. We have to override them to avoid
+    # trying to modify the readonly data and mask arrays
+    def __inop(self, other):
```
Comment (mhvk) [inline numpy/ma/core.py]: Single `_`?

File: numpy/ma/core.py [hunk 10]
```diff
@@ -6206,16 +6227,23 @@ def __str__(self):
...
 
     def __repr__(self):
-        return 'masked'
-
-    def flatten(self):
-        return masked_array([self._data], dtype=float, mask=[True])
+        if self is masked:
+            return 'masked'
+        else:
+            # something is wrong, make it obvious
```
Comment (mhvk) [inline numpy/ma/core.py]: Just to be sure, this will not happen in any normal usage, correct? (e.g., from the test below i assume `.copy` returns something else)

File: numpy/ma/core.py (hunk 1 above)
Comment (mhvk) [inline numpy/ma/core.py]: With my above suggestion, I'd hope this won't be necessary any more.

File: numpy/ma/tests/test_core.py [hunk 11]
```diff
@@ -4732,6 +4732,49 @@ def test_ctor(self):
...
+        assert_equal(x.mask, False)
+
+        # original should be unchanged
+        assert_equal(np.ma.masked.data, 0)
+        assert_equal(np.ma.masked.mask, True)
+
+    def test_no_copies(self):
+        MC = np.ma.core.MaskedConstant
+        assert_(not isinstance(np.ma.masked[...], MC))
+        assert_(not isinstance(np.ma.masked.copy(), MC))
```
Comment (mhvk) [inline numpy/ma/tests/test_core.py]: Can we add a test that checks the actual class that is returned by `[...]` and `.copy()`?

File: numpy/ma/tests/test_core.py [hunk 12]
```diff
@@ -4732,6 +4732,49 @@ def test_ctor(self):
...
+
+    def test_immutable(self):
+        assert_raises(ValueError, operator.setitem, np.ma.masked.data, (), 1)
```
Comment (mhvk) [inline numpy/ma/tests/test_core.py]: Also add the test for just trying to set `np.ma.masked` itself?

**Review by mhvk — COMMENTED** [review]
I have to admit I looked at this without really referring to the rest of the code, which means I didn't understand some things. Apologies! I think my suggestion for `__new__` and `__array_finalize__` would be an improvement, though.

File: numpy/ma/core.py (hunk 3 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: So as not to clash with any subclasses of `MaskConstant`, perhaps? Why not `__`?

File: numpy/ma/core.py (hunk 4 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Good plan

File: numpy/ma/core.py (hunk 8 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Inherited from the superclass

File: numpy/ma/core.py (hunk 10 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Well, that's the hope. There were many different ways that this could happen before, and this change made it easy to spot.

It also means that subclasses of `MaskedConstant` don't masquerade as `masked` either.

File: numpy/ma/tests/test_core.py (hunk 11 above)
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: Are you worried the type might be `ndarray` here?

File: numpy/ma/core.py (hunk 9 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Guess I named this to look like `__iadd__` without actually looking like a special method

File: numpy/ma/core.py (hunk 1 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Pretty sure we're stuck with this - too much of the `MaskedArray` code calls `.view(type(self))`, and we really don't want to have to special case every single one of those.

An alternative approach would be to allow `__array_finalize__` to have a return value, which replaces `self`

File: numpy/ma/core.py (hunk 3 above)
Comment (mhvk) [inline numpy/ma/core.py]: `__` to me is associated with `dunder` while this is really just a private attribute.

File: numpy/ma/core.py (hunk 10 above)
Comment (mhvk) [inline numpy/ma/core.py]: OK, makes sense.

File: numpy/ma/core.py (hunk 9 above)
Comment (mhvk) [inline numpy/ma/core.py]: Would prefer that, but `_iop` would also be fine

File: numpy/ma/tests/test_core.py (hunk 11 above)
Comment (mhvk) [inline numpy/ma/tests/test_core.py]: It is more that I had no idea what it would be! (As mentioned, I didn't look at the rest of the code, but from this PR it was not obvious what the result should be, and on current master it is a `MaskedConstant`)

File: numpy/ma/core.py (hunk 3 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: `__` is a special mechanism in python to do name-mangling, to make attributes unique to their class, so that property names don't clash with subclasses

File: numpy/ma/core.py (hunk 1 above)
Comment (mhvk) [inline numpy/ma/core.py]: It would seem `anything.view(MaskedConstant)` is almost certainly an error....  And setting `__class__` is quite awful. But I defer to your judgment - I'm on holidays and worry looking in more detail is not going to be a short process...

(and I agree we cannot change `__array_finalize__`)

Comment (eric-wieser) [inline numpy/ma/core.py]: > It would seem anything.view(MaskedConstant) is almost certainly an error

Written that way, yes. The problem is any kind of subclass-capable code that tries to reattach the subclass, perhaps after allocating a new buffer. That shouldn't have to know about `np.ma`, so `np.ma` needs to be able to make `.view(MaskedConstant)` do something sensible

File: numpy/ma/core.py (hunk 3 above)
Comment (mhvk) [inline numpy/ma/core.py]: Ah, learned something new! So, here it makes sense (but for `__inop` it then does not; presumably that *should* be inherited - not that anyone really should want to inherit `MaskedConstant`...)

Comment (eric-wieser) [inline numpy/ma/core.py]: Heck, I should probably add a `del __inop` there anyway, which makes the name irrelevant

File: numpy/ma/core.py (hunk 9 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: Actually calling it `__iop__`, so that `np.ma.masked.__add__.__name__` prints something somewhat reasonable

File: numpy/ma/tests/test_core.py (hunk 11 above)
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: Will adjust to asserting the direct type here

File: numpy/ma/core.py (hunk 6 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: We need the view casing anyway, because `MaskedArray.__new__` calls `.view`

File: numpy/ma/tests/test_core.py (hunk 11 above)
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: Fun fact - copying 0d masked arrays doesn't copy the mask

**Comment by eric-wieser** [comment]
Hopefully addressed most of the reviews. This raised #9430, but I'd rather open that can of worms in another PR.

File: numpy/ma/tests/test_core.py [hunk 13]
```diff
@@ -4732,6 +4732,56 @@ def test_ctor(self):
...
+        # when making a view or a copy, downcast to MaskedArray
+        MC = np.ma.core.MaskedConstant
+
+        m_sl = np.ma.masked[...]
+        assert_equal(type(m_sl), np.ma.MaskedArray)
+        assert_equal(m_sl.mask, True)
+
+        m_copy = np.ma.masked.copy()
+        assert_equal(type(m_copy), np.ma.MaskedArray)
+        # assert_equal(m_copy.mask, True)  - gh-9430
```
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: This can of worms left unopened

Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: Fixed here: #9432

File: numpy/ma/core.py [hunk 14]
```diff
@@ -6206,16 +6229,30 @@ def __str__(self):
...
     def __reduce__(self):
         """Override of MaskedArray's __reduce__.
         """
         return (self.__class__, ())
 
+    # inplace operations have no effect. We have to override them to avoid
+    # trying to modify the readonly data and mask arrays
+    def __iop__(self, other):
+        return self
+    __iadd__ = \
```
Comment (charris) [inline numpy/ma/core.py]: So this is OK because the singleton can never be unmasked? Might be cleaner stylewise to just assign the function rather than have all the line continuations.

Comment (eric-wieser) [inline numpy/ma/core.py]: Correct, if the singleton is unmasked there are much bigger problems.

File: numpy/ma/tests/test_core.py [hunk 15]
```diff
@@ -4732,6 +4732,56 @@ def test_ctor(self):
...
+
+        with BytesIO() as f:
+            pickle.dump(np.ma.masked, f)
+            f.seek(0)
+            res = pickle.load(f)
+        assert_(res is np.ma.masked)
+
+    def test_write_to_copy(self):
+        # gh-9328
+        x = np.ma.masked.copy()
```
Comment (charris) [inline numpy/ma/tests/test_core.py]: So `x` is a downcast masked array, not the singleton in this case?

Comment (charris) [inline numpy/ma/tests/test_core.py]: If so, strange behavior for `copy`.

Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: Correct, because that seemed the least bad option, with the others being:

* `return self` - this is clearly not a copy, and will fail when trying to write to it, whereas normally you would expect to be able to write to the result of copy() without side-effects
* `return MaskedConstant(self)` - this is super confusing, because it printed as `masked`, and would be neither constant nor singleton

Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: I take it back, this should `return self`, because `x.copy() is x` is already true for `np.True_`, so there's precedent for that behaviour for scalars

File: numpy/ma/core.py (hunk 14 above)
Comment (eric-wieser) [inline numpy/ma/core.py]: I think I prefer the continuations, as they ensure that `__iop__` doesn't get mispelt as one of the other methods

**Comment by eric-wieser** [comment]
Possibly an improvements release note?

File: numpy/ma/tests/test_core.py [hunk 16]
```diff
@@ -4732,6 +4732,43 @@ def test_ctor(self):
...
+    def test_copy(self):
+        # gh-9328
+        # copy is a no-op, like it is with np.True_
+        assert_equal(
+            np.ma.masked.copy() is np.ma.masked,
+            np.True_.copy() is np.True_)
+
+    def test_immutable(self):
+        orig = np.ma.masked
+        assert_raises(np.ma.core.MaskError, operator.setitem, orig, (), 1)
```
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: It's a bit weird how this raises a different error, but shouldn't be a big deal

Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: (and importantly, hasn't changed in this PR)

**Comment by charris** [comment]
If the behavior has changed, it should probably be in the compatibility notes.

**Comment by charris** [comment]
@eric-wieser Could you add a release note so I can merge this?

**Comment by eric-wieser** [comment]
Looks like I forgot to push the release notes fix. I'm still away from git for the next day or two

**Comment by ahaldane** [comment]
Code LGTM.

The following might be out of the scope of this PR, but I came across them while testing:

The singleton's shape can be edited:
```python
>>> np.ma.masked.shape = (1,1,1)   # 'ravel' does this too
>>> np.ma.array([1], mask=True)[0].shape
(1, 1, 1)
```

Singleton does not always behave like a 0d array:
```python
>>> np.ma.masked.mean()  # fails, unlike other 0d masked arrays
AttributeError: 'int' object has no attribute 'shape'
>>> np.ma.masked.filled()  # fails, unlike other 0d masked arrays
AttributeError: 'MaskedConstant' object has no attribute '_fill_value'
>>> x = m.repeat(4)
>>> x
masked
>>> x.shape
(4,)
```

These are all preexisting behavior, so maybe we can leave it. It's similar to what this PR fixes though.

**Review by ahaldane — APPROVED** [review]

**Comment by eric-wieser** [comment]
@charris - a summary of the behaviour changes. I think these are all bug fixes, making none of them release-note worthy:

* `masked` is better at being the singleton it is documented to be, by not allowing copies:
  * round-tripped through pickle by identity, rather than making a copy
  * `masked.copy()` doesn't copy, because it doesn't need to.
* The buffers owned by `masked` are readonly

Perhaps this latter one deserves a release note.

**Comment by eric-wieser** [comment]
Added a release note about point 2, @charris

**Comment by charris** [comment]
Thanks Eric.

**Comment by mhvk** [comment]
That `np.ma.masked` is no longer hashable is breaking tests in astropy - https://github.com/astropy/astropy/issues/6645 (I think we had discussion about this, but am not quite sure where...)

**Comment by eric-wieser** [comment]
Did this patch change whether `masked` could be `hash`ed?
